### PR TITLE
Initial partial implementation for MVD3 support

### PR DESCRIPTION
**Pull request by adevress**
- Add basic MVD3 support to Brion::circuit
- Add mvd-tool as a subproject for MVD3 parsing
- Move feature from Brain API to Brion API ( improve perfs with MVD3 )
- Add tests for MVD3 parsing
- Add Nix support

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -22,6 +22,8 @@ set(BRION_MAINTAINER
 set(BRION_LICENSE LGPL)
 
 set(COMMON_PROJECT_DOMAIN ch.epfl.bluebrain)
+set(MVD_TOOL_DISABLE_INSTALL TRUE) # enable MVD-tool embedded usage
+
```
Comment (eile) [inline CMakeLists.txt]: Not used!?

File: brion/CMakeLists.txt [hunk 2]
```diff
@@ -4,6 +4,8 @@
 # This file is part of Brion <https://github.com/BlueBrain/Brion>
 #
 
+include_directories(${CMAKE_CURRENT_SOURCE_DIR}/../mvd-tool/include/ ${CMAKE_CURRENT_SOURCE_DIR}/../mvd-tool/deps/highfive/include )
+
```
Comment (eile) [inline brion/CMakeLists.txt]: Should be provided through the target. @tribal-tec can you fix this?

File: brion/circuit.cpp [hunk 3]
```diff
@@ -153,22 +174,52 @@ class Circuit::Impl
...
         {
-        case NEURONCLASS_ETYPE:
-        {
+        case NEURONCLASS_ETYPE:{
             CircuitTable::const_iterator i = _table.find( SECTION_ETYPES );
             return i != _table.end() ? i->second : Strings();
         }
 
-        case NEURONCLASS_MTYPE:
+        case NEURONCLASS_MTYPE:{
```
Comment (eile) [inline brion/circuit.cpp]: ?

Also, can you try to keep with the coding style of the respective project in the future?

File: brion/circuit.cpp [hunk 4]
```diff
@@ -42,18 +45,29 @@ enum Section
     SECTION_UNKNOWN
 };
 
-class Circuit::Impl
+class Circuit::Impl{
+public:
+  virtual ~Impl(){}
+
+  virtual NeuronMatrix get( const GIDSet& gids, const uint32_t attributes ) const =0;
```
Comment (favreau) [inline brion/circuit.cpp]: Formatting details but choose between '=0' and '= 0'

File: brion/circuit.cpp [hunk 5]
```diff
@@ -69,24 +83,31 @@ class Circuit::Impl
...
-                current = it->second;
-            else
-                _table[current].push_back( content );
+                LookUp::const_iterator it = sections.find( content );
+                if( it != sections.end() )
+                    current = it->second;
+                else
+                    _table[current].push_back( content );
+            }
+        }catch(std::ifstream::failure & e){
```
Comment (favreau) [inline brion/circuit.cpp]: Formatting again catch( std::ifstream::failure& e )

File: brion/circuit.cpp [hunk 6]
```diff
@@ -202,13 +254,199 @@ class Circuit::Impl
...
+
+//////////////////////////////////////////////////
+/// Circuit Pimpl mapper
+//////////////////////////////////////////////////
+
+Circuit::Impl* Circuit::instanciate(const std::string & source){
+    namespace fs = boost::filesystem;
+    fs::path path = source;
+    const std::string ext = fs::extension( path );
+    if( ext == ".mvd" || ext == ".mvd2" ){
```
Comment (favreau) [inline brion/circuit.cpp]: extension is in uppercase? lowercase?

File: CMakeLists.txt (hunk 1 above)
Comment (adevress) [inline CMakeLists.txt]: It is used, the line just disable the installation of the header and the cmd line tool of MVDtool. It is a shared component, it should not be installed as part of Brion.

File: brion/circuit.cpp (hunk 6 above)
Comment (adevress) [inline brion/circuit.cpp]: It was like this before, I did not change it, just moved it :)

File: brion/circuit.cpp (hunk 3 above)
Comment (adevress) [inline brion/circuit.cpp]: np

File: brion/CMakeLists.txt (hunk 2 above)
Comment (adevress) [inline brion/CMakeLists.txt]: Both HighFive and mvd-tool are header only, there is no associated target.

File: .gitsubprojects [hunk 7]
```diff
@@ -2,3 +2,5 @@
 git_subproject(Servus https://github.com/HBPVIS/Servus.git 567f038)
 git_subproject(Lunchbox https://github.com/Eyescale/Lunchbox.git b41edcf)
 git_subproject(vmmlib https://github.com/Eyescale/vmmlib.git b41e1a5)
+git_subproject(mvd-tool ssh://bbpcode.epfl.ch/hpc/MVD-tool 517ca7f)
```
Comment (eile) [inline .gitsubprojects]: Will create a project on github. Can we get rid of the dash, e.g., MVDTool?

Comment (adevress) [inline .gitsubprojects]: Such a huge change, I'm not sure it will be possible :D 
no problem
Can you associate it with the HPC account of github ?

Comment (eile) [inline .gitsubprojects]: Done. HPC can write, you are admin.

**Comment by adevress** [comment]
Ok Guys,

- I switched the repo of MVD tool to the github one.
- Rotation with quaternion is now fully supported.

Ready to integrate for my side

File: .gitsubprojects [hunk 8]
```diff
@@ -2,3 +2,7 @@
 git_subproject(Servus https://github.com/HBPVIS/Servus.git 567f038)
 git_subproject(Lunchbox https://github.com/Eyescale/Lunchbox.git b41edcf)
 git_subproject(vmmlib https://github.com/Eyescale/vmmlib.git b41e1a5)
+git_subproject(mvd-tool https://github.com/BlueBrain/MVDTool.git 1d38937)
+
+
+
```
Comment (eile) [inline .gitsubprojects]: remove newlines

File: CMake/GitExternal.cmake [hunk 9]
```diff
@@ -118,12 +115,7 @@ function(GIT_EXTERNAL DIR REPO tag)
...
-      if(GIT_EXTERNAL_LOCAL_OPTIONAL)
-        message(STATUS "${DIR} clone failed: ${error}\n")
-        return()
-      else()
-        message(FATAL_ERROR "${DIR} clone failed: ${error}\n")
-      endif()
+      message(FATAL_ERROR "${DIR} clone failed: ${error}\n")
     endif()
 
     # checkout requested tag
```
Comment (eile) [inline CMake/GitExternal.cmake]: revert this file - you've got an old CMake/common?

File: CMakeLists.txt (hunk 1 above)
Comment (eile) [inline CMakeLists.txt]: A superproject install always installs all artefacts. For packages, superproject builds are not used but you depend on packages.

If we want to address this, we'll need to do it in one place for all non-public subprojects.

File: brion/CMakeLists.txt (hunk 2 above)
Comment (eile) [inline brion/CMakeLists.txt]: See MVDTool PR, and add below:
```
set(BRION_LINK_LIBRARIES PUBLIC Lunchbox Servus vmmlib
                         PRIVATE MVDTool ${Boost_LIBRARIES} ${HDF5_LIBRARIES}
                                 ${CMAKE_THREADS_LIB_INIT}
)
```

File: brion/circuit.cpp [hunk 10]
```diff
@@ -153,12 +178,79 @@ class Circuit::Impl
...
+                std::advance(it_gid, i);
+                LBWARN << "Error parsing circuit position or orientation for gid "
+                       << *it_gid << ". Morphology not transformed." << std::endl;
+            }
+        }
+        return positions;
+
+    }
+
+    virtual Vector4fs getQuatRotations( const GIDSet& gids ) const
```
Comment (eile) [inline brion/circuit.cpp]: use Quaternionf from vmmlib

File: brain/circuit.cpp [hunk 11]
```diff
@@ -141,63 +141,41 @@ neuron::Morphologies Circuit::loadMorphologies( const GIDSet& gids,
...
-        }
-        catch( const boost::bad_lexical_cast& )
-        {
-            LBWARN << "Error parsing circuit position or orientation for gid "
-                   << *gid << ". Morphology not transformed." << std::endl;
-            matrix = Matrix4f::IDENTITY;
-        }
-        #pragma omp critical (brain_circuit_getTransforms)
-        ++gid;
+
```
Comment (eile) [inline brain/circuit.cpp]: use Quaternionf

File: brion/circuit.h [hunk 12]
```diff
@@ -69,6 +69,23 @@ class Circuit : public boost::noncopyable
...
+     */
+    BRION_API Vector3fs getPositions( const GIDSet& gids ) const;
+
+    /**
+     * @param gids set of neurons of interest; if empty, all neurons in the
+     *             circuit file are considered
+     * @return Vector of the rotation quaternions for the given cells.
+     *
+     */
+    BRION_API Vector4fs getQuatRotations( const GIDSet& gids ) const;
```
Comment (eile) [inline brion/circuit.h]: ```Quaternionf getRotations(...```

**Comment by adevress** [comment]
Ok Done @eile 

- code has been updated to use vmmll::quaternion
- logic has been moved to brain::circuit instead of brion::circuit like requested.
- I integrated the thing with your BlueConfig interface, it is now one switch to move to MVD3
- added some new tests

However,
- Your modifications on the MVDTool cmake targets triggers an Ubuntu 14.04 Jenkins error
- The BBPTestData used on your jenkins plan need to be update from trunk. A new circuit file has been aded there.

**Comment by adevress** [comment]
Ok Guys
Problem related to BBPTestData and cmake 2 / common are gone. Have a look when you have time

**Comment by jafyvilla** [comment]
Apparently the installation folder of the TestData module had been removed. I triggered a new build and it seems the module was updated correctly this time.

**Comment by adevress** [comment]
Ok, just tell me or pass by if you have any comment on this

**Pull request merged**